README.md

```diff
@@ -1,7 +1,246 @@
+```
+
+The new type, `Repr`, is a singleton type that establishes a connection between
+a run-time value and a type.  When we pattern match on `IsEq`, we can simply
+inspect (i.e., pattern match on) the contained `Repr` value to determine the
+types of the sub-terms:
+
+```haskell
+withBoolExprs :: Expr tp -> a -> ([Expr 'EBool] -> a) -> a
+withBoolExprs e def k =
+  case e of
+    BoolLit {} -> k [e]
+    Lt {} -> k [e]
+    IsEq rep e1 e2 ->
+      case rep of
+        BoolRepr ->
+          -- Because we used a GADT pattern match, we know that tp ~ EBool
+          k [e, e1, e2]
+        IntRepr -> def
+    _ -> def
+```
+
+# Package Structure
+
+This package provides three main types of functionality:
+
+* Typeclasses mirroring core Haskell classes, but adapted to parameterized types
+* Data structures suitable for holding values of parameterized types
+* Utilities for working with parameterized types, including tools for proving properties at the type level (dependently-typed programming in Haskell)
+
+## Typeclasses
+
+* Data.Parameterized.Classes
+
+  This module contains a number of basic classes lifted to parameterized types,
+  including `EqF`, `OrdF`, `ShowF`, and `HashableF`.  It also re-exports
+  a few types from base that are useful for working with parameterized types,
+  including `TestEquality`.
+
+  The related module Data.Parameterized.TH.GADT provides Template Haskell
+  functions to automatically implement instances of some of these classes.
+
+* Data.Parameterized.ClassesC
+
+  This module defines classes like Data.Parameterized.Classes, except that the
+  class methods accept an additional parameter for comparing sub-terms.
+
+* Data.Parameterized.TraversableFC
+
+  This module generalizes `Functor`, `Foldable`, and `Traversable` to
+  parameterized types.  In these operations, type parameters must be preserved.
+
+* Data.Parameterized.TraversableF
+
+  This module is like Data.Parameterized.TraversableFC, but intended for types
+  that have a single parametric type parameter, rather than two.  The most
+  common use of these functions and classes is with the `MapF` type described
+  below.
+
+
+## Data Structures
+
+This package provides data structures that are either lifted to hold
+parameterized types or otherwise type indexed.  The following modules implement
+data structures:
+
+* Data.Parameterized.Context (`Assignment (f :: k -> Type) (ctx :: Ctx k)`)
+
+  `Assignment` is a sequence type that holds values of parameterized types.
+  It is essentially a snoc list (i.e., a list that is extended on the right
+  instead of the left).  In the default implementation, indexing is O(log(n))
+  time and total.
+
+* Data.Parameterized.List (`List (f :: k -> Type) [k]`)
+
+  `List` is the plain Haskell list lifted to hold values of parameterized
+  types.  Moreover, it uses the data kind lifted list syntax instead of the
+  `Ctx` type.  Indexing into `List` is total but O(n).
```

Comment:
The `Ctx` data structure is never described, just referred to here and used in the `Context` above.